crates/sui-core/src/in_mem_execution_cache.rs

```diff
@@ -0,0 +1,1242 @@
+                Ok(true)
+            }
+            _ => Ok(false),
+        }
+    }
+}
+
+pub trait ExecutionCacheWrite: Send + Sync {
+    /// Write the output of a transaction.
+    ///
+    /// Because of the child object consistency rule (readers that observe parents must observe all
+    /// children of that parent, up to the parent's version bound), implementations of this method
+    /// must not write any top-level (address-owned or shared) objects before they have written all
+    /// of the object-owned objects (i.e. child objects) in the `objects` list.
+    ///
+    /// In the future, we may modify this method to expose finer-grained information about
+    /// parent/child relationships. (This may be especially necessary for distributed object
+    /// storage, but is unlikely to be an issue before we tackle that problem).
+    ///
+    /// This function may evict the mutable input objects (and successfully received objects) of
+    /// transaction from the cache, since they cannot be read by any other transaction.
+    ///
+    /// Any write performed by this method immediately notifies any waiter that has previously
+    /// called notify_read_objects_for_execution or notify_read_objects_for_signing for the object
+    /// in question.
+    fn write_transaction_outputs(&self, epoch_id: EpochId, tx_outputs: TransactionOutputs);
+}
+
+enum ObjectEntry {
+    Object(Object),
+    Deleted,
+    Wrapped,
+}
+
+impl From<Object> for ObjectEntry {
+    fn from(object: Object) -> Self {
+        ObjectEntry::Object(object)
+    }
+}
+
+type MarkerKey = (EpochId, ObjectID);
+
+enum CacheResult<T> {
+    /// Entry is in the cache
+    Hit(T),
+    /// Entry is not in the cache and is known to not exist
+    NegativeHit,
+    /// Entry is not in the cache and may or may not exist in the store
+    Miss,
+}
+
+pub struct InMemoryCache {
+    /// The object dirty set. All writes go into this table first. After we flush the data to the
+    /// db, the data is removed from this table and inserted into the object_cache.
+    ///
+    /// This table may contain both live and dead objects, since we flush both live and dead
+    /// objects to the db in order to support past object queries on fullnodes.
+    /// When we move data into the object_cache we only retain the live objects.
+    ///
+    /// Further, we only remove objects in FIFO order, which ensures that the the cached
+    /// sequence of objects has no gaps. In other words, if we have versions 4, 8, 13 of
+    /// an object, we can deduce that version 9 does not exist. This also makes child object
+    /// reads efficient. `object_cache` cannot contain a more recent version of an object than
+    /// `objects`, and neither can have any gaps. Therefore if there is any object <= the version
+    /// bound for a child read in objects, it is the correct object to return.
+    objects: DashMap<ObjectID, BTreeMap<SequenceNumber, ObjectEntry>>,
```

Comment:
Probably need better naming for clarity.
One suggestion is to call "objects" as "all_dirty_objects", "objects_cache" as "cached_move_objects", and "packages" as "cached_move_packages".

Reply:
I also recommend partition the fields into two separate structs, one contain uncommitted results, and one contain cached data from DB.